src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java

```diff
@@ -0,0 +1,514 @@
+  private static final Logging logger = new Logging(ParquetRowBuffer.class);
+
+  private static final Set<ColumnPhysicalType> TIME_SUPPORTED_PHYSICAL_TYPES =
+      new HashSet<>(Arrays.asList(ColumnPhysicalType.SB4, ColumnPhysicalType.SB8));
+  private static final Set<ColumnPhysicalType> TIMESTAMP_SUPPORTED_PHYSICAL_TYPES =
+      new HashSet<>(Arrays.asList(ColumnPhysicalType.SB8, ColumnPhysicalType.SB16));
+
+  private final Map<String, Pair<ColumnMetadata, Integer>> fieldIndex;
+  private final Map<String, String> metadata;
+  private final List<List<Object>> data;
+  private final List<List<Object>> tempData;
+
+  private MessageType schema;
+
+  ParquetRowBuffer(
+      SnowflakeStreamingIngestChannelInternal<ParquetRowBuffer.ParquetChunkData> channel) {
+    super(channel);
+    fieldIndex = new HashMap<>();
+    metadata = new HashMap<>();
+    data = new ArrayList<>();
+    tempData = new ArrayList<>();
+  }
+
+  @Override
+  public void setupSchema(List<ColumnMetadata> columns) {
+    fieldIndex.clear();
+    metadata.clear();
+    List<Type> parquetTypes = new ArrayList<>();
+    int id = 1;
+    for (ColumnMetadata column : columns) {
+      Type type = buildField(column, id);
+      parquetTypes.add(type);
+      fieldIndex.put(column.getName(), new Pair<>(column, parquetTypes.size() - 1));
+      if (!column.getNullable()) {
+        this.nonNullableFieldNames.add(column.getName());
+      }
+      this.statsMap.put(column.getName(), new RowBufferStats(column.getCollation()));
+
+      if (this.owningChannel.getOnErrorOption() == OpenChannelRequest.OnErrorOption.ABORT) {
+        this.tempStatsMap.put(column.getName(), new RowBufferStats(column.getCollation()));
+      }
+
+      id++;
+    }
+    schema = new MessageType("bdec", parquetTypes);
+  }
+
+  private Type buildField(ColumnMetadata column, int id) {
+    Type parquetType;
+    String name = column.getName();
+
+    ColumnPhysicalType physicalType;
+    ColumnLogicalType logicalType;
+    try {
+      physicalType = ColumnPhysicalType.valueOf(column.getPhysicalType());
+      logicalType = ColumnLogicalType.valueOf(column.getLogicalType());
+    } catch (IllegalArgumentException e) {
+      throw new SFException(
+          ErrorCode.UNKNOWN_DATA_TYPE, column.getLogicalType(), column.getPhysicalType());
+    }
+
+    this.metadata.put(
+        Integer.toString(id), logicalType.getOrdinal() + "," + physicalType.getOrdinal());
+
+    LogicalTypeAnnotation.DecimalLogicalTypeAnnotation decimalLogicalTypeAnnotation =
+        column.getScale() != null && column.getPrecision() != null
+            ? LogicalTypeAnnotation.DecimalLogicalTypeAnnotation.decimalType(
+                column.getScale(), column.getPrecision())
+            : null;
+
+    // Handle differently depends on the column logical and physical types
+    Type.Repetition repetition =
+        column.getNullable() ? Type.Repetition.OPTIONAL : Type.Repetition.REQUIRED;
```

Comment:
For my understanding: for what do we need the repetion?

Reply:
Parquet `Type.Repetition` in general supports repeated values for the same row column, like a list of values.
This generator uses only
- either 0 or 1 value for nullable data type (`OPTIONAL`: 0 or none value if it is null)
- or exactly 1 value for non-nullable data type (`REQUIRED`)
Thanks for asking! I added all this to comments